chain-signatures/node/src/protocol/posit.rs

```
@@ -0,0 +1,177 @@
+                    tracing::warn!(
+                        ?id,
+                        "received a protocol posit for an id that we already proposed"
+                    );
+                    return PositInternalAction::Reply(PositAction::Reject);
+                }
+
+                // No further checks necessary, we can just accept the posit.
+                PositInternalAction::Reply(PositAction::Accept)
+            }
+            PositAction::Start(participants) => {
+                if self.posits.contains_key(&id) {
+                    tracing::warn!(?id, "received an invalid Start for a protocol we proposed");
+                    return PositInternalAction::Reply(PositAction::Reject);
+                }
+
+                PositInternalAction::StartProtocol(
+                    participants.to_vec(),
+                    Positor::Deliberator(from),
+                )
+            }
+            PositAction::Accept | PositAction::Reject => {
+                let mut entry = match self.posits.entry(id) {
+                    Entry::Occupied(counter) => counter,
+                    Entry::Vacant(_) => {
+                        tracing::warn!(
+                            ?id,
+                            ?action,
+                            "received an action for a protocol we did NOT propose"
+                        );
+                        return PositInternalAction::None;
+                    }
+                };
+
+                let counter = entry.get_mut();
+                if action.is_accept() {
+                    counter.accepts.insert(from);
+                } else {
+                    counter.rejects.insert(from);
+                }
+
+                let enough_votes = counter.accepts.len() >= threshold
+                    && counter.accepts.len() + counter.rejects.len() == counter.participants.len();
+                if !enough_votes {
+                    return PositInternalAction::None;
+                }
```

Could we also check if we have too many rejections already? Maybe not in this PR but it would be nice to fail early if we know we can't find enough peers with the necessary data still available.

---

Yep, that's possible, but requires adding back the abort I got rid of. Not a bad thing to do, since it would be nice to abort the ongoing protocol, and have the proposer restart it

---

Added the check but will not abort the protocol for now. Left it as a TODO in the future